addons/survey/controllers/main.py

```diff
@@ -732,35 +727,130 @@ def _get_user_input_domain(self, survey, line_filter_domain, **post):
-                elif answer_id:
-                    line_choices.append(answer_id)
-                    answers = request.env['survey.question.answer'].browse([answer_id])
-                if answer_id:
-                    question_id = answers[0].matrix_question_id or answers[0].question_id
-                    search_filters.append({
-                        'row_id': row_id,
-                        'answer_id': answer_id,
-                        'question': question_id.title,
-                        'answers': '%s%s' % (answers[0].value, ': %s' % answers[1].value if len(answers) > 1 else '')
-                    })
-        if line_choices:
-            line_filter_domain = expression.AND([[('suggested_answer_id', 'in', line_choices)], line_filter_domain])
 
-        user_input_domain = self._get_user_input_domain(survey, line_filter_domain, **post)
-        user_input_lines = request.env['survey.user_input'].sudo().search(user_input_domain).mapped('user_input_line_ids')
+        answer_cols_to_rows_mapping, answer_user_input_line_ids = self._get_filters_from_post(post)
+
+        # Matrix, Multiple choice, Simple choice filters
+        if answer_cols_to_rows_mapping:
+            ans_ids, row_ids = [], []
+            for filter_ans_id, filter_row_ids in answer_cols_to_rows_mapping.items():
+                ans_ids.append(filter_ans_id)
+                row_ids += filter_row_ids
+
+            answers = request.env['survey.question.answer'].browse(list(filter(bool, ans_ids+row_ids)))  # exclude '0'
+            # For performance, caches all useful columns of survey_question_answer in first query
+            answers_only = answers.filtered(lambda a: not a.matrix_question_id)   # not row_ids
+
+            for answer in answers_only:
+                for row_id in answer_cols_to_rows_mapping[answer.id]:
+                    filters_domains += self._prepare_matching_answer_domain(answer.question_type, answer, row_id=row_id)
+                    search_filters.append(self._prepare_search_filter('A', answer, row_id=row_id))
+
+        # Char_box, Text_box, Numerical_box, Date, Datetime filters
+        if answer_user_input_line_ids:
+            user_input_line_ids = request.env['survey.user_input.line'].browse(answer_user_input_line_ids)
+            for input_line_id in user_input_line_ids:
+                value, operator = self._get_value_with_operator(input_line_id)
+                filters_domains += self._prepare_matching_answer_domain(input_line_id.answer_type, input_line_id, operator=operator, value=value)
+                search_filters.append(self._prepare_search_filter('L', input_line_id, value=value))
+
+        user_input_domain = self._get_results_page_user_input_domain(survey, **post)
+        user_input_lines = self._get_matching_user_input_lines(filters_domains, user_input_domain)
 
         return user_input_lines, search_filters
+
+    def _get_matching_user_input_lines(self, filters_domains, user_input_domain):
```

Comment:
We tend to avoid making methods that are only called once (unless they are meant to be overridden elsewhere).

It makes the code harder to read as you have to follow multiple "paths" instead of just reading the function.

By "inlined", I mean that the function code could be "inlined" here, which is removing the function and copying the code basically. Sorry if that was unclear ^^